### PR TITLE
Pass in bucked id to smartdexingstep to fix compatibility with redex

### PR DESCRIPTION
**Pull request by kageiit**
Follow up on https://github.com/facebook/buck/pull/2198#discussion_r258121644

### CODE REVIEW

**Comment by kageiit** [comment]
cc @styurin

File: src/com/facebook/buck/android/SmartDexingStep.java [hunk 1]
```diff
@@ -463,9 +471,20 @@ static void createDxStepForDxPseudoRule(
...
+      boolean useDexBuckedId) {
 
+    Optional<String> buckedId = Optional.empty();
     String output = outputPath.toString();
+    if (useDexBuckedId) {
+      // We know what the output file name is ("classes.dex" or "classesN.dex") as these
+      // are generated in SplitZipStep and passed around as part of a multi-map - it is
+      // simply easier and cleaner to extract the dex file number to be used as unique
+      // identifier rather than creating another map and pass it around
+      String[] tokens = outputPath.getFileName().toString().split("\\.")[0].split("classes");
```
Comment (styurin) [inline src/com/facebook/buck/android/SmartDexingStep.java]: You can use `Files.getNameWithoutExtension()` instead.

Comment (kageiit) [reply src/com/facebook/buck/android/SmartDexingStep.java]: Done

**Comment by styurin** [comment]
Build is broken

**Comment by kageiit** [comment]
> Build is broken

huh it was working before recent change iirc. let me take a look!

**Comment by kageiit** [comment]
@styurin fixed the build and rebased again. Sorry about that! Feel free to merge once build passes

**Comment by kageiit** [comment]
Seems like the new logic using files api is broken in integration tests. Will fix and push up

**Comment by kageiit** [comment]
Fixed the bug and pushed. Please re-import :)

**Comment by kageiit** [comment]
Hmm seems like travis has surfaced new failures. Im unable to run these integration tests locally. Ill continue to iterate and ping again once travis is fully green. Sorry for the trouble @styurin

File: src/com/facebook/buck/android/SmartDexingStep.java [hunk 2]
```diff
@@ -463,9 +472,20 @@ static void createDxStepForDxPseudoRule(
...
 
+    Optional<String> buckedId = Optional.empty();
     String output = outputPath.toString();
+    if (useDexBuckedId) {
+      // We know what the output file name is ("classes.dex" or "classesN.dex") as these
+      // are generated in SplitZipStep and passed around as part of a multi-map - it is
+      // simply easier and cleaner to extract the dex file number to be used as unique
+      // identifier rather than creating another map and pass it around
+      String[] tokens = Files.getNameWithoutExtension(outputPath.toString()).split("classes");
+      String id = tokens.length == 1 ? "" /* primary */ : tokens[1] /* secondary */;
```
Comment (styurin) [inline src/com/facebook/buck/android/SmartDexingStep.java]: This needs to have "tokens.length == 0", the same as it was in the original change.

Comment (kageiit) [reply src/com/facebook/buck/android/SmartDexingStep.java]: interesting. ok will make the change to see if the tests pass finally :)

Comment (kageiit) [reply src/com/facebook/buck/android/SmartDexingStep.java]: Done and pushed. Lets wait for CI :)

**Comment by kageiit** [comment]
Seems like 2 tests failed. lot better than before, Are you planning to fix these internally @styurin ?

**Comment by styurin** [comment]
I tried to debug the failed test, it looks like this PR is not complete. I'd recommend going through the failed test and fix the PR.

**Comment by kageiit** [comment]
@styurin I cant seem to run this test locally. Complaints about missing ndk platform toolchains. Whats the best way to set this up on a mac so i can reproduce. Its not clear why the build fails with proguard on and exit code 10

**Comment by kageiit** [comment]
@styurin CI is now green. The only failure is from `BuildEndToEndTest` which seems to be flaky

**Pull request closed without merging**